File: mo-system/mo-system-service/src/main/java/ink/ckx/mo/system/service/impl/SysDeptServiceImpl.java
package ink.ckx.mo.system.service.impl;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.collection.CollUtil;
import cn.hutool.core.util.StrUtil;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import ink.ckx.mo.common.web.enums.StatusEnum;
import ink.ckx.mo.common.web.model.Option;
import ink.ckx.mo.system.api.constant.SystemConstant;
import ink.ckx.mo.system.api.model.entity.SysDept;
import ink.ckx.mo.system.api.model.form.DeptForm;
import ink.ckx.mo.system.api.model.query.DeptListQuery;
import ink.ckx.mo.system.api.model.vo.dept.DeptDetailVO;
import ink.ckx.mo.system.api.model.vo.dept.DeptListVO;
import ink.ckx.mo.system.converter.DeptConverter;
import ink.ckx.mo.system.mapper.SysDeptMapper;
import ink.ckx.mo.system.service.SysDeptService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * 部门业务实现类
 *
 * @author chenkaixin
 */
@Service
@RequiredArgsConstructor
public class SysDeptServiceImpl extends ServiceImpl<SysDeptMapper, SysDept> implements SysDeptService {

    private final DeptConverter deptConverter;

    @Override
    public List<DeptListVO> listDepartments(DeptListQuery deptListQuery) {
        // 查询参数
        String keywords = deptListQuery.getKeywords();
        StatusEnum status = deptListQuery.getStatus();

        // 查询数据
        List<SysDept> deptList = this.list(new LambdaQueryWrapper<SysDept>()
                .select(SysDept::getId, SysDept::getParentId, SysDept::getName, SysDept::getStatus,
                        SysDept::getSort, SysDept::getCreateTime, SysDept::getUpdateTime)
                .eq(status != null, SysDept::getStatus, status)
                .like(StrUtil.isNotBlank(keywords), SysDept::getName, keywords)
                .orderByAsc(SysDept::getSort));

        List<DeptListVO> list = new ArrayList<>();
        if (CollUtil.isNotEmpty(deptList)) {
            Set<Long> cacheDeptIds = deptList.stream().map(SysDept::getId).collect(Collectors.toSet());
            for (SysDept dept : deptList) {
                Long parentId = dept.getParentId();
                // 不在缓存ID列表的parentId是顶级节点ID，以此作为递归开始
                if (!cacheDeptIds.contains(parentId)) {
                    list.addAll(this.recurDepartments(parentId, deptList));
                    // 避免重复递归
                    cacheDeptIds.add(parentId);
                }
            }
        }

        //  列表为空说明所有的节点都是独立的
        if (list.isEmpty()) {
            return deptList.stream().map(item -> {
                DeptListVO deptListVO = new DeptListVO();
                BeanUtil.copyProperties(item, deptListVO);
                return deptListVO;
            }).collect(Collectors.toList());
        }

        return list;
    }

    /**
     * 递归生成部门层级列表
     *
     * @param parentId
     * @param deptList
     * @return
     */
    private List<DeptListVO> recurDepartments(Long parentId, List<SysDept> deptList) {
        return deptList.stream().filter(dept -> dept.getParentId().equals(parentId)).map(dept -> {
            DeptListVO deptListVO = deptConverter.entity2VO(dept);
            List<DeptListVO> children = this.recurDepartments(dept.getId(), deptList);
            deptListVO.setChildren(children);
            return deptListVO;
        }).toList();
    }

    @Override
    public DeptDetailVO getDeptDetail(Long deptId) {
        SysDept entity = this.getOne(new LambdaQueryWrapper<SysDept>()
                .select(SysDept::getId, SysDept::getName, SysDept::getParentId, SysDept::getStatus, SysDept::getSort)
                .eq(SysDept::getId, deptId));

        return deptConverter.entity2DetailVO(entity);
    }

    @Override
    public List<Option<Long>> listDeptOptions() {
        List<SysDept> deptList = this.list(new LambdaQueryWrapper<SysDept>()
                .select(SysDept::getId, SysDept::getParentId, SysDept::getName)
                .eq(SysDept::getStatus, StatusEnum.ENABLE)
                .orderByAsc(SysDept::getSort));

        //   List<Option> options = recurDeptTreeOptions(SystemConstants.ROOT_NODE_ID, deptList);
        return this.buildDeptTree(deptList);
    }

    @Override
    public Long saveDept(DeptForm deptForm) {
        SysDept sysDept = deptConverter.form2Entity(deptForm);
        // 部门路径
        String treePath = this.generateDeptTreePath(deptForm.getParentId());
        sysDept.setTreePath(treePath);
        // 保存部门并返回部门ID
        this.save(sysDept);
        return sysDept.getId();
    }

    @Override
    public void updateDept(Long deptId, DeptForm formData) {
        // form->entity
        SysDept sysDept = deptConverter.form2Entity(formData);
        sysDept.setId(deptId);
        // 部门路径
        String treePath = this.generateDeptTreePath(formData.getParentId());
        sysDept.setTreePath(treePath);
        // 更新部门
        this.updateById(sysDept);
    }

    @Override
    public void deleteByIds(String ids) {
        List<Long> idList = StrUtil.split(ids, ',', 0, true, Long::valueOf);
        if (CollUtil.isNotEmpty(idList)) {
            // 删除部门及子部门
            idList.forEach(deptId ->
                    this.remove(new LambdaQueryWrapper<SysDept>()
                            .eq(SysDept::getId, deptId)
                            .or()
                            .apply("concat (',',tree_path,',') like concat('%,',{0},',%')", deptId)));
        }
    }

    /**
     * 递归生成部门表格层级列表
     *
     * @param deptList
     * @return
     */
    private List<Option<Long>> buildDeptTree(List<SysDept> deptList) {
        if (CollUtil.isEmpty(deptList)) {
            return new ArrayList<>();
        }
        List<Option<Long>> returnList = new ArrayList<>();
        List<Long> tempList = deptList.stream().map(SysDept::getId).toList();
        for (SysDept dept : deptList) {
            // 如果是顶级节点, 遍历该父节点的所有子节点
            if (!tempList.contains(dept.getParentId())) {
                Option<Long> option = new Option<>(dept.getId(), dept.getName());
                this.recursionFn(deptList, option);
                returnList.add(option);
            }
        }
        if (CollUtil.isEmpty(returnList)) {
            deptList.forEach(dept -> {
                Option<Long> option = new Option<>(dept.getId(), dept.getName());
                returnList.add(option);
            });
        }
        return returnList;
    }

    /**
     * 递归列表
     */
    private void recursionFn(List<SysDept> list, Option<Long> t) {
        // 得到子节点列表
        List<Option<Long>> childList = getChildList(list, t);
        t.setChildren(childList);
        for (Option<Long> tChild : childList) {
            if (this.hasChild(list, tChild)) {
                this.recursionFn(list, tChild);
            }
        }
    }

    /**
     * 得到子节点列表
     */
    private List<Option<Long>> getChildList(List<SysDept> list, Option<Long> t) {
        List<Option<Long>> tlist = new ArrayList<>();
        for (SysDept n : list) {
            if (n.getParentId() != null && Objects.equals(n.getParentId(), t.getValue())) {
                Option<Long> option = new Option<>(n.getId(), n.getName());
                tlist.add(option);
            }
        }
        return tlist;
    }

    /**
     * 判断是否有子节点
     */
    private boolean hasChild(List<SysDept> list, Option<Long> t) {
        return this.getChildList(list, t).size() > 0;
    }

    /**
     * 部门路径生成
     *
     * @param parentId
     * @return
     */
    private String generateDeptTreePath(Long parentId) {
        String treePath = null;
        if (SystemConstant.ROOT_ID.equals(parentId)) {
            treePath = parentId + "";
        } else {
            SysDept parent = this.getOne(new LambdaQueryWrapper<SysDept>()
                    .select(SysDept::getId, SysDept::getTreePath)
                    .eq(SysDept::getId, parentId));
            if (parent != null) {
                treePath = parent.getTreePath() + "," + parent.getId();
            }
        }
        return treePath;
    }
}

File: mo-admin/mo-admin-service/src/main/kotlin/ink/ckx/mo/admin/service/impl/SysDeptServiceImpl.kt
package ink.ckx.mo.admin.service.impl

import cn.hutool.core.bean.BeanUtil
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import ink.ckx.mo.admin.api.constant.AdminConstant
import ink.ckx.mo.admin.api.model.entity.SysDept
import ink.ckx.mo.admin.api.model.form.DeptForm
import ink.ckx.mo.admin.api.model.query.DeptListQuery
import ink.ckx.mo.admin.api.model.vo.dept.DeptDetailVO
import ink.ckx.mo.admin.api.model.vo.dept.DeptListVO
import ink.ckx.mo.admin.converter.DeptConverter
import ink.ckx.mo.admin.mapper.SysDeptMapper
import ink.ckx.mo.admin.service.SysDeptService
import ink.ckx.mo.common.web.enums.StatusEnum
import ink.ckx.mo.common.web.model.Option
import org.springframework.stereotype.Service

/**
 * 部门业务实现类
 *
 * @author chenkaixin
 */
@Service
class SysDeptServiceImpl(
    private val deptConverter: DeptConverter
) : ServiceImpl<SysDeptMapper, SysDept>(), SysDeptService {

    override fun listDepartments(deptListQuery: DeptListQuery): List<DeptListVO> {
        // 查询参数
        val keywords: String? = deptListQuery.keywords
        val status: StatusEnum? = deptListQuery.status

        // 查询数据
        val deptList = ktQuery()
            .select(
                SysDept::id, SysDept::parentId, SysDept::name, SysDept::status,
                SysDept::sort, SysDept::createTime, SysDept::updateTime
            )
            .eq(status != null, SysDept::status, status)
            .like(!keywords.isNullOrBlank(), SysDept::name, keywords)
            .orderByDesc(SysDept::sort, SysDept::updateTime)
            .list()
        val resultList: MutableList<DeptListVO> = mutableListOf()
        if (deptList.isNotEmpty()) {
            val cacheDeptIds = deptList.map { e -> e.id }.toMutableSet()
            for (dept in deptList) {
                val parentId = dept.parentId
                // 不在缓存ID列表的parentId是顶级节点ID，以此作为递归开始
                if (parentId !in cacheDeptIds) {
                    resultList.addAll(recurDepartments(parentId, deptList))
                    // 避免重复递归
                    cacheDeptIds.add(parentId)
                }
            }
        }

        //  列表为空说明所有的节点都是独立的
        return resultList.takeIf { it.isNotEmpty() } ?: deptList.map {
            val deptListVO = DeptListVO()
            BeanUtil.copyProperties(it, deptListVO)
            deptListVO
        }
    }

    /**
     * 递归生成部门层级列表
     *
     * @param parentId
     * @param deptList
     * @return
     */
    private fun recurDepartments(parentId: Long?, deptList: List<SysDept>): List<DeptListVO> {
        return deptList
            .filter { it.parentId == parentId }
            .map {
                val deptListVO = deptConverter.entity2VO(it)
                val children = recurDepartments(it.id, deptList)
                deptListVO.children = children
                deptListVO
            }
            .toList()
    }

    override fun getDeptDetail(deptId: Long): DeptDetailVO? {
        val entity = ktQuery()
            .select(SysDept::id, SysDept::name, SysDept::parentId, SysDept::status, SysDept::sort)
            .eq(SysDept::id, deptId)
            .one()
        return deptConverter.entity2DetailVO(entity)
    }

    override fun listDeptOptions(): List<Option<Long>> {
        val deptList = ktQuery()
            .select(SysDept::id, SysDept::parentId, SysDept::name)
            .eq(SysDept::status, StatusEnum.ENABLE)
            .orderByAsc(SysDept::sort)
            .list()
        return buildDeptTree(deptList)
    }

    override fun saveDept(deptForm: DeptForm): Long? {
        val sysDept = deptConverter.form2Entity(deptForm)
        // 部门路径
        val treePath = generateDeptTreePath(deptForm.parentId)
        sysDept.treePath = treePath
        // 保存部门并返回部门ID
        save(sysDept)
        return sysDept.id
    }

    override fun updateDept(deptId: Long, deptForm: DeptForm) {
        // form->entity
        val sysDept = deptConverter.form2Entity(deptForm)
        sysDept.id = deptId
        // 部门路径
        val treePath = generateDeptTreePath(deptForm.parentId)
        sysDept.treePath = treePath
        // 更新部门
        updateById(sysDept)
    }

    override fun deleteByIds(ids: String) {
        val idList = ids.split(',').mapNotNull { it.toLongOrNull() }
        if (idList.isNotEmpty()) {
            // 删除部门及子部门
            idList.forEach {
                ktUpdate()
                    .eq(SysDept::id, it)
                    .or()
                    .apply("concat (',',tree_path,',') like concat('%,',{0},',%')", it)
                    .remove()
            }
        }
    }

    /**
     * 递归生成部门表格层级列表
     *
     * @param deptList
     * @return
     */
    private fun buildDeptTree(deptList: List<SysDept>): List<Option<Long>> {
        if (deptList.isEmpty()) {
            return mutableListOf()
        }
        val returnList: MutableList<Option<Long>> = mutableListOf()
        val tempList = deptList.map(SysDept::id).toList()
        for (dept in deptList) {
            // 如果是顶级节点, 遍历该父节点的所有子节点
            if (dept.parentId !in tempList) {
                val option = Option(dept.id, dept.name)
                recursionFn(deptList, option)
                returnList.add(option)
            }
        }
        if (returnList.isEmpty()) {
            deptList.forEach {
                val option = Option(it.id, it.name)
                returnList.add(option)
            }
        }
        return returnList
    }

    /**
     * 递归列表
     */
    private fun recursionFn(list: List<SysDept>, t: Option<Long>) {
        // 得到子节点列表
        val childList = getChildList(list, t)
        t.children = childList
        for (tChild in childList) {
            if (hasChild(list, tChild)) {
                recursionFn(list, tChild)
            }
        }
    }

    /**
     * 得到子节点列表
     */
    private fun getChildList(list: List<SysDept>, t: Option<Long>): List<Option<Long>> {
        val childList: MutableList<Option<Long>> = mutableListOf()
        for (n in list) {
            if (n.parentId == t.value) {
                val option: Option<Long> = Option(n.id, n.name)
                childList.add(option)
            }
        }
        return childList
    }

    /**
     * 判断是否有子节点
     */
    private fun hasChild(list: List<SysDept>, t: Option<Long>): Boolean = getChildList(list, t).isNotEmpty()

    /**
     * 部门路径生成
     *
     * @param parentId
     * @return
     */
    private fun generateDeptTreePath(parentId: Long?): String {
        var treePath = ""
        if (AdminConstant.ROOT_ID == parentId) {
            treePath = parentId.toString() + ""
        } else {
            val parent = ktQuery()
                .select(SysDept::id, SysDept::treePath)
                .eq(SysDept::id, parentId)
                .one()
            if (parent != null) {
                treePath = parent.treePath + "," + parent.id
            }
        }
        return treePath
    }
}